File: go/net/manager.go
package net

import (
	"errors"
	"sync"
	"time"

	"github.com/vibing/zgrnet/noise"
)

// SessionManager manages multiple sessions with different peers.
// It provides lookup by local index and by remote public key.
type SessionManager struct {
	mu sync.RWMutex

	// Sessions indexed by local index
	byIndex map[uint32]*noise.Session

	// Sessions indexed by remote public key
	byPubkey map[noise.PublicKey]*noise.Session

	// Index allocator
	nextIndex uint32
}

// NewSessionManager creates a new session manager.
func NewSessionManager() *SessionManager {
	return &SessionManager{
		byIndex:   make(map[uint32]*noise.Session),
		byPubkey:  make(map[noise.PublicKey]*noise.Session),
		nextIndex: 1, // Start from 1, 0 is reserved
	}
}

// CreateSession creates and registers a new session.
// If a session already exists for the given public key, it is replaced.
func (m *SessionManager) CreateSession(remotePK noise.PublicKey, sendKey, recvKey noise.Key) (*noise.Session, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Generate a unique local index
	localIndex := m.allocateIndex()
	if localIndex == 0 {
		return nil, ErrNoFreeIndex
	}

	// Create the session
	session, err := noise.NewSession(noise.SessionConfig{
		LocalIndex: localIndex,
		SendKey:    sendKey,
		RecvKey:    recvKey,
		RemotePK:   remotePK,
	})
	if err != nil {
		return nil, err
	}

	// Remove any existing session for this peer
	if existing, ok := m.byPubkey[remotePK]; ok {
		delete(m.byIndex, existing.LocalIndex())
	}

	// Register the new session
	m.byIndex[localIndex] = session
	m.byPubkey[remotePK] = session

	return session, nil
}

// RegisterSession registers an externally created session.
// This is useful when the handshake is performed separately.
func (m *SessionManager) RegisterSession(session *noise.Session) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	localIndex := session.LocalIndex()
	remotePK := session.RemotePublicKey()

	// Check for index collision
	if _, exists := m.byIndex[localIndex]; exists {
		return ErrIndexInUse
	}

	// Remove any existing session for this peer
	if existing, ok := m.byPubkey[remotePK]; ok {
		delete(m.byIndex, existing.LocalIndex())
	}

	// Register
	m.byIndex[localIndex] = session
	m.byPubkey[remotePK] = session

	return nil
}

// GetByIndex retrieves a session by its local index.
// Returns nil if not found.
func (m *SessionManager) GetByIndex(index uint32) *noise.Session {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.byIndex[index]
}

// GetByPubkey retrieves a session by the remote peer's public key.
// Returns nil if not found.
func (m *SessionManager) GetByPubkey(pk noise.PublicKey) *noise.Session {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.byPubkey[pk]
}

// RemoveSession removes a session by its local index.
func (m *SessionManager) RemoveSession(index uint32) {
	m.mu.Lock()
	defer m.mu.Unlock()

	session, ok := m.byIndex[index]
	if !ok {
		return
	}

	delete(m.byIndex, index)
	delete(m.byPubkey, session.RemotePublicKey())
}

// RemoveByPubkey removes a session by the remote peer's public key.
func (m *SessionManager) RemoveByPubkey(pk noise.PublicKey) {
	m.mu.Lock()
	defer m.mu.Unlock()

	session, ok := m.byPubkey[pk]
	if !ok {
		return
	}

	delete(m.byIndex, session.LocalIndex())
	delete(m.byPubkey, pk)
}

// ExpireSessions removes all expired sessions.
// Returns the number of sessions removed.
func (m *SessionManager) ExpireSessions() int {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Store session pointers directly to avoid second lookup
	var expired []*noise.Session
	for _, session := range m.byIndex {
		if session.IsExpired() {
			expired = append(expired, session)
		}
	}

	for _, session := range expired {
		delete(m.byIndex, session.LocalIndex())
		delete(m.byPubkey, session.RemotePublicKey())
	}

	return len(expired)
}

// Count returns the number of active sessions.
func (m *SessionManager) Count() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.byIndex)
}

// Sessions returns a snapshot of all sessions.
// The returned slice is safe to iterate while the manager is modified.
func (m *SessionManager) Sessions() []*noise.Session {
	m.mu.RLock()
	defer m.mu.RUnlock()

	sessions := make([]*noise.Session, 0, len(m.byIndex))
	for _, session := range m.byIndex {
		sessions = append(sessions, session)
	}
	return sessions
}

// ForEach iterates over all sessions with a callback.
// The callback should not modify the manager.
func (m *SessionManager) ForEach(fn func(*noise.Session)) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	for _, session := range m.byIndex {
		fn(session)
	}
}

// Clear removes all sessions.
func (m *SessionManager) Clear() {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.byIndex = make(map[uint32]*noise.Session)
	m.byPubkey = make(map[noise.PublicKey]*noise.Session)
}

// allocateIndex generates a unique session index.
// Must be called with mu held.
// Returns 0 if no free index is available (extremely unlikely).
func (m *SessionManager) allocateIndex() uint32 {
	startIndex := m.nextIndex
	for {
		index := m.nextIndex
		m.nextIndex++
		if m.nextIndex == 0 {
			m.nextIndex = 1 // Wrap around, skip 0
		}

		// Check if index is in use
		if _, exists := m.byIndex[index]; !exists {
			return index
		}

		// Check if we've wrapped around completely
		if m.nextIndex == startIndex {
			return 0 // No free index available
		}
	}
}

// StartExpiryWorker starts a background goroutine that periodically
// removes expired sessions. Returns a stop function.
func (m *SessionManager) StartExpiryWorker(interval time.Duration) func() {
	stop := make(chan struct{})
	done := make(chan struct{})

	go func() {
		defer close(done)
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for {
			select {
			case <-ticker.C:
				m.ExpireSessions()
			case <-stop:
				return
			}
		}
	}()

	return func() {
		close(stop)
		<-done
	}
}

// Errors for SessionManager
var (
	ErrIndexInUse  = errors.New("net: session index already in use")
	ErrNoFreeIndex = errors.New("net: no free session index available")
)

File: zig/src/net/manager.zig
//! Session manager for multiple peers.

const std = @import("std");
const Mutex = std.Thread.Mutex;
const Allocator = std.mem.Allocator;

const noise = @import("mod.zig");

pub const Key = noise.Key;
pub const Session = noise.Session;
pub const SessionConfig = noise.SessionConfig;

/// Manager errors.
pub const ManagerError = error{
    IndexInUse,
    OutOfMemory,
    NoFreeIndex,
};

/// Manages multiple sessions with different peers.
pub const SessionManager = struct {
    allocator: Allocator,
    mutex: Mutex = .{},
    by_index: std.AutoHashMap(u32, *Session),
    by_pubkey: std.AutoHashMap(Key, *Session),
    next_index: u32 = 1,

    /// Creates a new session manager.
    pub fn init(allocator: Allocator) SessionManager {
        return .{
            .allocator = allocator,
            .by_index = std.AutoHashMap(u32, *Session).init(allocator),
            .by_pubkey = std.AutoHashMap(Key, *Session).init(allocator),
        };
    }

    /// Deinitializes the manager and frees all sessions.
    pub fn deinit(self: *SessionManager) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        var it = self.by_index.valueIterator();
        while (it.next()) |session_ptr| {
            self.allocator.destroy(session_ptr.*);
        }

        self.by_index.deinit();
        self.by_pubkey.deinit();
    }

    /// Creates and registers a new session.
    pub fn createSession(
        self: *SessionManager,
        remote_pk: Key,
        send_key: Key,
        recv_key: Key,
    ) ManagerError!*Session {
        self.mutex.lock();
        defer self.mutex.unlock();

        const local_index = self.allocateIndex() orelse return ManagerError.NoFreeIndex;

        const session = self.allocator.create(Session) catch return ManagerError.OutOfMemory;
        session.* = Session.init(.{
            .local_index = local_index,
            .send_key = send_key,
            .recv_key = recv_key,
            .remote_pk = remote_pk,
        });

        // Remove existing session for this peer
        if (self.by_pubkey.fetchRemove(remote_pk)) |removed| {
            _ = self.by_index.remove(removed.value.local_index);
            self.allocator.destroy(removed.value);
        }

        self.by_index.put(local_index, session) catch {
            self.allocator.destroy(session);
            return ManagerError.OutOfMemory;
        };

        self.by_pubkey.put(remote_pk, session) catch {
            _ = self.by_index.remove(local_index);
            self.allocator.destroy(session);
            return ManagerError.OutOfMemory;
        };

        return session;
    }

    /// Registers an externally created session.
    pub fn registerSession(self: *SessionManager, session: *Session) ManagerError!void {
        self.mutex.lock();
        defer self.mutex.unlock();

        const local_index = session.local_index;
        const remote_pk = session.remote_pk;

        if (self.by_index.contains(local_index)) {
            return ManagerError.IndexInUse;
        }

        // Remove existing session for this peer
        if (self.by_pubkey.fetchRemove(remote_pk)) |removed| {
            _ = self.by_index.remove(removed.value.local_index);
            self.allocator.destroy(removed.value);
        }

        self.by_index.put(local_index, session) catch return ManagerError.OutOfMemory;
        self.by_pubkey.put(remote_pk, session) catch {
            _ = self.by_index.remove(local_index);
            return ManagerError.OutOfMemory;
        };
    }

    /// Gets a session by local index.
    pub fn getByIndex(self: *SessionManager, index: u32) ?*Session {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.by_index.get(index);
    }

    /// Gets a session by remote public key.
    pub fn getByPubkey(self: *SessionManager, pk: Key) ?*Session {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.by_pubkey.get(pk);
    }

    /// Removes a session by local index.
    pub fn removeSession(self: *SessionManager, index: u32) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        if (self.by_index.get(index)) |session| {
            _ = self.by_pubkey.remove(session.remote_pk);
            _ = self.by_index.remove(index);
            self.allocator.destroy(session);
        }
    }

    /// Removes a session by remote public key.
    pub fn removeByPubkey(self: *SessionManager, pk: Key) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        if (self.by_pubkey.get(pk)) |session| {
            _ = self.by_index.remove(session.local_index);
            _ = self.by_pubkey.remove(pk);
            self.allocator.destroy(session);
        }
    }

    /// Removes all expired sessions.
    /// Returns the number of sessions removed.
    pub fn expireSessions(self: *SessionManager) usize {
        self.mutex.lock();
        defer self.mutex.unlock();

        var expired = std.ArrayListUnmanaged(u32){};
        defer expired.deinit(self.allocator);

        var it = self.by_index.iterator();
        while (it.next()) |entry| {
            if (entry.value_ptr.*.isExpired()) {
                expired.append(self.allocator, entry.key_ptr.*) catch continue;
            }
        }

        for (expired.items) |index| {
            if (self.by_index.get(index)) |session| {
                _ = self.by_pubkey.remove(session.remote_pk);
                _ = self.by_index.remove(index);
                self.allocator.destroy(session);
            }
        }

        return expired.items.len;
    }

    /// Returns the number of sessions.
    pub fn count(self: *SessionManager) usize {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.by_index.count();
    }

    /// Clears all sessions.
    pub fn clear(self: *SessionManager) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        var it = self.by_index.valueIterator();
        while (it.next()) |session_ptr| {
            self.allocator.destroy(session_ptr.*);
        }

        self.by_index.clearAndFree();
        self.by_pubkey.clearAndFree();
    }

    fn allocateIndex(self: *SessionManager) ?u32 {
        const start_index = self.next_index;
        while (true) {
            const index = self.next_index;
            self.next_index +%= 1;
            if (self.next_index == 0) {
                self.next_index = 1;
            }

            if (!self.by_index.contains(index)) {
                return index;
            }

            // Check if we've wrapped around completely
            if (self.next_index == start_index) {
                return null; // No free index available
            }
        }
    }
};

// Tests
const testing = std.testing;

test "create session" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk = Key.fromBytes([_]u8{1} ** 32);
    const session = try m.createSession(pk, Key.zero, Key.zero);

    try testing.expect(session.local_index > 0);
    try testing.expectEqual(@as(usize, 1), m.count());
}

test "get by index" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk = Key.fromBytes([_]u8{1} ** 32);
    const session = try m.createSession(pk, Key.zero, Key.zero);
    const index = session.local_index;

    try testing.expect(m.getByIndex(index) != null);
    try testing.expect(m.getByIndex(99999) == null);
}

test "get by pubkey" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk = Key.fromBytes([_]u8{1} ** 32);
    _ = try m.createSession(pk, Key.zero, Key.zero);

    try testing.expect(m.getByPubkey(pk) != null);
    try testing.expect(m.getByPubkey(Key.fromBytes([_]u8{9} ** 32)) == null);
}

test "remove session" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk = Key.fromBytes([_]u8{1} ** 32);
    const session = try m.createSession(pk, Key.zero, Key.zero);
    const index = session.local_index;

    m.removeSession(index);

    try testing.expect(m.getByIndex(index) == null);
    try testing.expect(m.getByPubkey(pk) == null);
    try testing.expectEqual(@as(usize, 0), m.count());
}

test "remove by pubkey" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk = Key.fromBytes([_]u8{1} ** 32);
    const session = try m.createSession(pk, Key.zero, Key.zero);
    const index = session.local_index;

    m.removeByPubkey(pk);

    try testing.expect(m.getByIndex(index) == null);
    try testing.expect(m.getByPubkey(pk) == null);
}

test "replace existing" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk = Key.fromBytes([_]u8{1} ** 32);

    const s1 = try m.createSession(pk, Key.zero, Key.zero);
    const idx1 = s1.local_index;

    const s2 = try m.createSession(pk, Key.fromBytes([_]u8{1} ** 32), Key.zero);
    const idx2 = s2.local_index;

    try testing.expect(m.getByIndex(idx1) == null);
    try testing.expect(m.getByIndex(idx2) != null);
    try testing.expectEqual(@as(usize, 1), m.count());
}

test "multiple peers" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    var sessions: [5]*Session = undefined;
    var pks: [5]Key = undefined;

    for (0..5) |i| {
        pks[i] = Key.fromBytes([_]u8{@intCast(i)} ** 32);
        sessions[i] = try m.createSession(pks[i], Key.zero, Key.zero);
    }

    try testing.expectEqual(@as(usize, 5), m.count());

    for (0..5) |i| {
        try testing.expect(m.getByPubkey(pks[i]) != null);
        try testing.expect(m.getByIndex(sessions[i].local_index) != null);
    }
}

test "expire sessions" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk1 = Key.fromBytes([_]u8{1} ** 32);
    const pk2 = Key.fromBytes([_]u8{2} ** 32);

    const s1 = try m.createSession(pk1, Key.zero, Key.zero);
    _ = try m.createSession(pk2, Key.zero, Key.zero);

    s1.expire();

    const removed = m.expireSessions();
    try testing.expectEqual(@as(usize, 1), removed);

    try testing.expect(m.getByPubkey(pk1) == null);
    try testing.expect(m.getByPubkey(pk2) != null);
}

test "clear" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    for (0..5) |i| {
        const pk = Key.fromBytes([_]u8{@intCast(i)} ** 32);
        _ = try m.createSession(pk, Key.zero, Key.zero);
    }

    m.clear();
    try testing.expectEqual(@as(usize, 0), m.count());
}

test "register session" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const session = try testing.allocator.create(Session);
    session.* = Session.init(.{
        .local_index = 12345,
        .send_key = Key.zero,
        .recv_key = Key.zero,
        .remote_pk = Key.fromBytes([_]u8{1} ** 32),
    });

    try m.registerSession(session);

    try testing.expect(m.getByIndex(12345) != null);
}

test "register session index collision" {
    var m = SessionManager.init(testing.allocator);
    defer m.deinit();

    const pk1 = Key.fromBytes([_]u8{1} ** 32);
    const s1 = try m.createSession(pk1, Key.zero, Key.zero);

    const s2 = try testing.allocator.create(Session);
    s2.* = Session.init(.{
        .local_index = s1.local_index,
        .send_key = Key.zero,
        .recv_key = Key.zero,
        .remote_pk = Key.fromBytes([_]u8{2} ** 32),
    });

    try testing.expectError(ManagerError.IndexInUse, m.registerSession(s2));
    testing.allocator.destroy(s2);
}
